### PR TITLE
[Package Resolver] Calculate concrete type abilities

### PR DESCRIPTION
**Pull request by amnn**
## Description

Expose a way for the package resolver to return the abilities for a concrete type (which in the case of generic types can depend on the abilities of their type parameters).

This will be used to expose abilities on types in GraphQL.

## Test Plan

New unit tests:

```
sui-package-resolver$ cargo nextest run
```

### CODE REVIEW

**Review by tnowacki — COMMENTED** [review]
Ability logic looks good to me. I'm glad `AbilitySet::polymorphic_abilities` is still proving useful enough. 

The only thing I might mention is do we care about perf here? Very  large types can be expensive to check for abilities (given the recursive nature and the O(type-tags) check at each level)

**Comment by amnn** [comment]
> The only thing I might mention is do we care about perf here?

Yeah, I think this is a reasonable question -- I'm not too concerned about queries through this API that are being called on actual objects on-chain (because they will cost gas to create and have limits imposed on them already), but we do also expose an API that allows people to construct arbitrary type instantiations and ask questions about them (like what their layout is, and what their abilities are), and we should put limits on those.

It's also probably worth making the traversal slightly cleverer at least in the case of abilities, to avoid fetching information for (actual) phantom type parameters, because we don't care about them, and in the case of on-chain objects it keeps the cost proportional to the size of the object on-chain and therefore its storage cost.  We unfortunately can't do the same for layout calculations, because we still need to calculate its type (including translating package IDs to their defining IDs), so I'll have to think about something else there -- probably an overall limit on the `ResolutionContext` size.

File: crates/sui-package-resolver/src/lib.rs [hunk 1]
```diff
@@ -1499,6 +1568,172 @@ mod tests {
...
+
+        // We would not be able to get an instance of this type, but in case the question is asked,
+        // its abilities would be empty.
+        let a3 = resolver
+            .abilities(type_("0xd0::m::O<0xd0::m::R, u64>"))
+            .await
+            .unwrap();
+        assert_eq!(a3, S::EMPTY);
+
+        // Key does not propagate up by itself, so this type is also uninhabitable.
```
Comment (wlmyng) [inline crates/sui-package-resolver/src/lib.rs:1678]: because P does not have Store?

**Review by wlmyng — APPROVED** [review]

**Comment by amnn** [comment]
I'll be adding the PRs to impose limits on type layout calculation as a stack on top of this one, meanwhile, I'll start staging later PRs to land into this one, as this is the current base!

**Pull request merged**